### DISCUSSION TITLE
Rust/C++ Interop

### DISCUSSION THREAD

**Discussion by benwis** [wasm-bindgen/wasm-bindgen#3893] [Q&A]
If I have a Rust project with C/C++ bindings, say treesitter and its parsers, and try to compile that together with cc to Webassembly with wasm32-unknown-unknown, will the ABI issues mentioned here cause errors between the two?
https://github.com/rust-lang/lang-team/blob/master/design-meeting-minutes/2021-04-21-wasm-abi.md

I've found plenty of references to it:
https://github.com/WebAssembly/tool-conventions/issues/88#issuecomment-544611586
https://github.com/rust-lang/rust/issues/63626
https://github.com/rust-lang/rust/blob/2da942f32802c8233a09744024dfbc34431adf65/src/librustc_target/abi/call/mod.rs#L570-L572
https://github.com/rust-lang/rust/issues/71871
https://github.com/rust-lang/rust/blob/9ab0749ce3fe33fdf84ac2f5b6c6e56b78dc10e8/compiler/rustc_target/src/spec/wasm32_unknown_unknown.rs#L20-L28

I'm trying to understand the problem space. Is it common/prevalent/limited to certain situations? Does it only matter if I compile something Rust to wasm32-wasi or something C to emscripten and then call it from Rust compiled to Webassembly?

**Comment by daxpedda** [comment]
The issue arises when 

> If I have a Rust project with C/C++ bindings, say treesitter and its parsers, and try to compile that together with cc to Webassembly with wasm32-unknown-unknown, will the ABI issues mentioned here cause errors between the two?

Yes, because whatever compiler `cc` is going to use is probably spec compliant, unlike Rustc.

> I'm trying to understand the problem space. Is it common/prevalent/limited to certain situations?

https://github.com/rust-lang/rust/issues/71871 explains it quite well, if you manage to avoid calling/linking or fix functions with this kind of ABI incompatibility, you can get away with it.

> Does it only matter if I compile something Rust to wasm32-wasi or something C to emscripten and then call it from Rust compiled to Webassembly?

Only Rustc targeting `wasm32-unknown-unknown` will produce non-spec-compliant ABI, as long as that is in the mix the issue might arise.

**Reply by benwis** [reply]
Thanks for the confirmation. Crossing my fingers this flag makes it into nightly soon(and wasm-bindgen supports it) to make this possible again: https://github.com/rust-lang/rust/pull/117919

**Reply by rafaelbeckel** [reply]
@benwis, the flag is now available on nightly, and the latest version of `wasm-bindgen` should support it. I put together a [minimal example](https://github.com/rafaelbeckel/test-c-rust-wasm), and now I'm closing all the tabs that led me to it with this comment.
